keyboards/handwired/lagrange/transport.c

```diff
@@ -0,0 +1,167 @@
+/* Copyright 2020 Dimitris Papavasiliou <dpapavas@protonmail.ch>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#undef NO_DEBUG
+
+#include <print.h>
+
+#include "quantum.h"
+#include "split_util.h"
+#include "timer.h"
+
+#include "lagrange.h"
+
+struct led_context {
+    led_t led_state;
+    layer_state_t layer_state;
+};
+
+uint8_t transceive(uint8_t b) {
+    for (SPDR = b ; !(SPSR & _BV(SPIF)) ; );
```

Comment:
Would it be possible to use the `spi_master` driver instead?

Reply:
Not without incurring a performance penalty, since the SPI master driver is designed for half-duplex operation and I transport data both ways (to notify the slave about LED and layer state).  It would also introduce an asymmetry in the implementation, as the slave would still need to use the low-level code.

I discussed introducing the SPI-based transport into QMK on Discord, in which case I intended to also update the SPI master driver and add a SPI client driver, but it was considered premature, as this will be the only SPI-based split keyboard in QMK.  Since there's also no need for full-duplex SPI operation, or a SPI client driver currently, introducing this functionality into QMK would probably also not be a good idea, hence the current implementation.

Reply:
Not sure what you mean by that, `spi_write()` will both send a byte and return the received byte at the same time: https://github.com/qmk/qmk_firmware/blob/master/drivers/avr/spi_master.c#L126

I actually did run into a problem where the ChibiOS driver was not doing this (#11404), but it has since been fixed, and this is an AVR board anyway.

Reply:
I had missed the fact that the single-byte write call does work full-duplex, because I was mostly looking at the multi-byte functions, `spi_transmit` and `spi_receive`, which do not.  So, to revise my statement above, I could use the SPI master driver, but I would need to retain almost all the code in `transport.c` anyway, as it's needed by the client.  From a quick look, I think I would only be able to remove the handful of setup calls in `transport_master_init`.

I  don't think the benefits would be worth the resulting asymmetry in the code, with separate setup code, separately defined SPI pins, etc. for the client and master side, but if required, I could make the changes.

Reply:
I don't know what you mean by asymmetry. As in there is no `spi_slave` driver? That's not really relevant, the idea is to make the code as platform-agnostic as possible.

Reply:
Yes.  The master part will use the abstracted code, the slave part will use the platform-dependent code.  On the whole, the code is exactly as platform-dependent as before:  It won't work if you configure it to use another AVR chip for instance, let alone another platform altogether, although half of it would, in theory, have worked.

Again, I don't mean to argue.  I'm only trying to explain why I found it preferable not to use the SPI master driver.  If you want me to use it, I'll try to make the necessary changes.   It doesn't really make a practical difference, so it's a matter of taste as far as I'm concerned.

Reply:
I've changed as much as possible to use the SPI driver.  Let me know if this version is acceptable.